File: src/Array/FindDuplicate.java
// https://leetcode.com/problems/find-the-duplicate-number/

package Array;

// Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.
// There is only one repeated number in nums, return this repeated number.
// You must solve the problem without modifying the array nums and uses only constant extra space.

import java.util.HashSet;

public class FindDuplicate {

    // Approach 1 - HashSet
    // Traverse the array and add each element to the HashSet using the add() method.
    // add() returns true if the element was not present previously and was added.
    // add() returns false if the element was present previously and hence was not added.
    // If add() returns false, that is our duplicate element.
    // Time Complexity - O(n)
    // Space Complexity - O(n)

    public static int findDuplicateUsingHashSet(int[] nums) {
        HashSet<Integer> hs = new HashSet<>();
        for (int num : nums) {
            if (!hs.add(num)) {
                return num;
            }
        }
        return -1;
    }

    // Approach 2 - Mark visited value within the array
    // Since all the values in the array are between 1 to n, and the array size is n+1
    // For each element at index i, set nums[element] to its negative value, ie. (nums[element] = -nums[element])
    // Whenever you encounter a negative number while accessing nums[element], return element
    // Time Complexity - O(n)
    // Space Complexity - O(1)

    public static int findDuplicateUsingMarking(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            int element = Math.abs(nums[i]);  // Prevents ArrayIndexOutOfBoundsException (accessing negative index)
            if (nums[element] < 0) {
                return element;
            }
            nums[element] = -nums[element];
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(findDuplicateUsingHashSet(new int[]{1, 3, 4, 2, 2}));  // 2
        System.out.println(findDuplicateUsingMarking(new int[]{1, 3, 4, 2, 2}));  // 2
    }
}

File: Love Babbar 450 DSA Questions/src/Cpp/Arrays/FindDuplicate.cpp
// https://leetcode.com/problems/find-the-duplicate-number/

// Reference (For Floyd cycle detection algo)
// https://www.youtube.com/watch?v=PvrxZaH_eZ4

#include <iostream>
#include <vector>
#include <cstdlib> // For abs()

// Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.
// There is only one repeated number in nums, return this repeated number.
// You must solve the problem without modifying the array nums and uses only constant extra space.

using namespace std;

// Approach 1 - Negative Marking (Modifies array, however we can restore the array back to the original one)
// Since the array has numbers in the range 1 to n only, and the size of the array is n + 1 we can use that to our benefit.
// For every number arr[i], set the value of mod(arr[i]) to -arr[i]
// If after setting mod(arr[i]) to -arr[i], the element arr[i] is positive, then the duplicate number is arr[i]
// Time Complexity: O(N)
// Space Complexity: O(1)
// Note: Even though this approach modifies the array, we can make all the digits positive and restore the original array.
int findDuplicateUsingNegativeMarking(vector<int> &nums)
{
    for (int i = 0; i < nums.size(); i++)
    {
        int index = abs(nums[i]);
        if (nums[index] < 0)
        {
            return index;
        }
        nums[index] *= -1;
    }
    return -1;
}

// Approach 2 - Floyd's tortoise and hare algo (cycle detection)
// Time Complexity : O(N)
// Space Complexity: O(1)
// Array is not modified
// Visualise a Linked List with each element arr[i] pointing to the node with index i
// Since one number is duplicate, 2 elements will point to the same node
// So we will have a cycle in the LL. That's why we can use Floyd's tortoise and hare algo to detect that cycle
// Take 2 pointers slow and fast initialise them to 0.
// while (true)
//   move slow by 1 and move fast by 2 ( slow = arr[slow] and fast = arr[arr[fast]])
// Once they meet, break the loop, set slow to 0, and increment both slow and fast by one this time
// Return the index of the point where they meet again. This is where the cycle starts.

int findDuplicateUsingTortoiseAndHare(vector<int> &nums)
{
    int slow{0};
    int fast{0};
    while (true)
    {
        slow = nums[slow];
        fast = nums[nums[fast]];
        if (slow == fast)
        {
            break;
        }
    }
    slow = 0;

    // Move both by 1
    while (true)
    {
        if (slow == fast)
        {
            return slow;
        }
        slow = nums[slow];
        fast = nums[fast];
    }
}

int main()
{
    vector<int> v1{3, 1, 3, 4, 2};
    vector<int> v2{3, 1, 3, 4, 2};
    cout << findDuplicateUsingNegativeMarking(v1) << '\n'; // 3
    cout << findDuplicateUsingTortoiseAndHare(v2) << '\n'; // 3
    return 0;
}